[jsk_recognition_utils] Enable publishing coefficients array

iory opened the pull request:
# What is this?

Added option to publish coefficients array to `poygon_array_publisher` node. If the polygon is a plane, the coeffiecients array is output in the same way as the output of plane detection.

708yamaguchi reviewed (commented):
I have a question about coef value.

When I use sample_polygon_array_publisher.launch, I got the following output.
```
$ roslaunch jsk_recognition_utils sample_polygon_array_publisher.launch
```
![Screenshot 2021-12-21 16:41:51](https://user-images.githubusercontent.com/19769486/146890865-7d287e95-482c-4fc9-a5de-2a81ff00b85e.png)
These places can be represented as the following:
- 0\*x + 0\*y + 1*z + 0 = 0
- 0\*x + 0\*y + 1*z + -0.5 = 0
- 0\*x + 0\*y + 1*z + -1 = 0
- 0\*x + -1\*y + 0*z + 1 = 0

So, coefficients [a, b c, d] should be 
- [0, 0, 1, 0]
- [0, 0, 1, -0.5]
- [0, 0, 1, -1]
- [0, -1, 0, 1]

In my understanding, [a, b, c] vector is the same as plane direction (arrow in RViz)


However, I got the following `/polygon_array_publisher/output/coefficients` value.
Is this correct?
```
 rostopic echo /polygon_array_publisher/output/coefficients                                         
header:                                                                                              
  seq: 1                                                                                             
  stamp:                                                                                             
    secs: 1640072311                                                                                 
    nsecs: 461468935                                                                                 
  frame_id: "base_link"                                                                              
coefficients:                                                                                        
  -                                                                                                  
    header:                                                                                          
      seq: 0                                                                                         
      stamp:                                                                                         
        secs: 1640072311                                                                             
        nsecs: 461468935                                                                             
      frame_id: "base_link"                                                                          
    values: [0.0, 0.0, -1.0, -0.0]                                                                   
  -                                                                                                  
    header:                                                                                          
      seq: 0                                                                                         
      stamp:                                                                                         
        secs: 1640072311                                                                             
        nsecs: 461468935                                                                             
      frame_id: "base_link"                                                                          
    values: [0.0, 0.0, -1.0, 0.5]                                                                    
  -                                                                                                  
    header:                                                                                          
      seq: 0                                                                                         
      stamp:                                                                                         
        secs: 1640072311                                                                             
        nsecs: 461468935                                                                             
      frame_id: "base_link"                                                                          
    values: [0.0, 0.0, -1.0, 1.0]                                                                    
  -                                                                                                  
    header:                                                                                          
      seq: 0                                                                                         
      stamp:                                                                                         
        secs: 1640072311                                                                             
        nsecs: 461468935                                                                             
      frame_id: "base_link"                                                                          
    values: [0.0, 1.0, 0.0, -1.0]                                                                    
---                                        
```

708yamaguchi commented on jsk_recognition_utils/node_scripts/polygon_array_publisher.py:
```diff
@@ -1,21 +1,60 @@
+from pcl_msgs.msg import ModelCoefficients
+
+
+def is_planar(points):
```
It would be helpful if you could write a summary or comment about the algorithm.

For example, I can easily understand this web page.
http://www.not-enough.org/abe/manual/argo/pointOnPlane.html

iory replied:
I see. Why don't you write a comment for understanding?

708yamaguchi replied:
OK, I will try.

708yamaguchi replied:
I created PR
https://github.com/iory/jsk_recognition/pull/13

iory commented:
> In my understanding, [a, b, c] vector is the same as plane direction (arrow in RViz). (I am not confident because I cannot find any documentation about plane direction)

Directions of arrows depend on visualization program. (not using coefficients)
jsk_rviz_plugin's visualization program internally calls jsk_recognition_utils's Plane class.

https://github.com/jsk-ros-pkg/jsk_recognition/blob/master/jsk_recognition_utils/src/geo/polygon.cpp#L87-L89
https://github.com/jsk-ros-pkg/jsk_recognition/blob/master/jsk_recognition_utils/src/geo/plane.cpp#L66-L70

```
 Plane((vertices[1] - vertices[0]).cross(vertices[2] - vertices[0]).normalized(), vertices[0]),
```

The direction of the normal is defined by each program, so it cannot be determined in general.

708yamaguchi commented:
> The direction of the normal is defined by each program, so it cannot be determined in general.

Thank you for your kind explanation.
I understand.

708yamaguchi commented:
I found that tests depending on `polygon_array_publisher.py` fails.
https://github.com/jsk-ros-pkg/jsk_recognition/runs/4602816415?check_suite_focus=true

This is because 
the number of `is_planar`'s return value is expected two
```
planar, coeffs = is_planar(polygon["points"])
```
but sometimes `is_planar` returns single value.
```
    # 3 points are always in a plane.
    if len(points) < 4:
        return True
    # 2d points are always in a plane.
    if len(points[0]) < 3:
        return True
```

```
$ rostest jsk_perception polygon_array_to_label_image.test                                                                                                                                                  
... logging to /home/leus/.ros/log/rostest-leus-ThinkPad-P1-Gen-3-5841.log                                                                                                                                  
[ROSUNIT] Outputting test results to /home/leus/.ros/test_results/jsk_perception/rostest-test_polygon_array_to_label_image.xml                                                                              
[ERROR] [1640177716.496418]: [[1.5, -1.0, 2.0], [-1.0, -1.0, 4.0], [0.5, 1.0, 3.0]]
[ERROR] [1640177716.497147]: True
Traceback (most recent call last):
  File "/home/leus/ros/melodic/src/jsk-ros-pkg/jsk_recognition/jsk_recognition_utils/node_scripts/polygon_array_publisher.py", line 126, in <module>
    p = PolygonArrayPublisher()
  File "/home/leus/ros/melodic/src/jsk-ros-pkg/jsk_recognition/jsk_recognition_utils/node_scripts/polygon_array_publisher.py", line 53, in __init__
    self.msg, self.coeffs_array_msg = self.parse_params(param, frame_id)
  File "/home/leus/ros/melodic/src/jsk-ros-pkg/jsk_recognition/jsk_recognition_utils/node_scripts/polygon_array_publisher.py", line 99, in parse_params
    planar, coeffs = is_planar(polygon["points"])
TypeError: 'bool' object is not iterable
^C^C^C[Testcase: testtest_polygon_array_to_label_image] ... ok
```

708yamaguchi reviewed (changes requested):
Sorry, the previous comment is request change.

iory commented:
Thanks.
I fixed the returning value.
https://github.com/jsk-ros-pkg/jsk_recognition/pull/2646/commits/14ab5ed3c99661332d67b691d7332d683e30d72f

I have also addressed the error when the input polygons are points or lines.
https://github.com/jsk-ros-pkg/jsk_recognition/pull/2646/commits/12ad4d4ac99f7783167ad917a7cfa2e0d26a2914
